eval/tricks.rkt

```diff
@@ -76,199 +79,311 @@
 ; Function calculates an ampl factor per input for a certain output and inputs using condition formulas,
 ;   where an ampl is an additional precision that needs to be added to srcs evaluation so,
 ;   that the output will be fixed in its precision when evaluating again
-(define (get-ampls op z srcs)
+; Additionaly, the function retures lower bound of ampl factor for the early exit mechanism
+; Output: '( '(upper-ampl-bound lower-ampl-bound) ...) with len(srcs) number of elements
+(define (get-bounds op z srcs)
   (case (object-name op)
     [(ival-mult)
-     ; k = 1: logspan(y)
-     ; k = 2: logspan(x)
+     ; Γ[*]'x     = 1
+     ; ↑ampl[*]'x = logspan(y)
+     ; ↓ampl[*]'x = 0
+     ; --------------
+     ; Γ[*]'y     = 1
+     ; ↑ampl[*]'y = logspan(x)
+     ; ↓ampl[*]'y = 0
      (define x (first srcs))
      (define y (second srcs))
-     (list (logspan y) ; exponent per x
-           (logspan x))] ; exponent per y
+     (list (cons (logspan y) 0) ; bounds per x
+           (cons (logspan x) 0))] ; bounds per y
 
     [(ival-div)
-     ; k = 1: logspan(y)
-     ; k = 2: logspan(x) + 2 * logspan(y)
+     ; Γ[/]'x     = 1
+     ; ↑ampl[/]'x = logspan(y)
+     ; ↓ampl[/]'x = 0
+     ; --------------
+     ; Γ[/]'y     = 1
+     ; ↑ampl[/]'y = logspan(x) + 2 * logspan(y)
+     ; ↓ampl[/]'y = 0
      (define x (first srcs))
      (define y (second srcs))
-     (list (logspan y) ; exponent per x
-           (+ (logspan x) (* 2 (logspan y))))] ; exponent per y
+     (list (cons (logspan y) 0) ; bounds per x
+           (cons (+ (logspan x) (* 2 (logspan y))) 0))] ; bounds per y
 
     [(ival-sqrt ival-cbrt)
-     ; sqrt: logspan(x)/2 - 1
-     ; cbrt: logspan(x)*2/3 - 1
+     ; Γ[sqrt]'x     = 0.5
+     ; ↑ampl[sqrt]'x = logspan(x)/2 - 1
+     ; ↓ampl[sqrt]'x = 0
+     ; --------------
+     ; Γ[cbrt]'x     = 1/3
+     ; ↑ampl[cbrt]'x = logspan(x)*2/3 - 1
+     ; ↓ampl[cbrt]'x = 0
      (define x (first srcs))
-     (list (quotient (logspan x) 2))]
+     (list (cons (quotient (logspan x) 2) 0))]
 
     [(ival-add ival-sub)
-     ; k = 1: maxlog(x) - minlog(z)
-     ; k = 2: maxlog(y) - minlog(z)
+     ; Γ[+ & -]'x     = |x/(x+y)| & |x/(x-y)|
+     ; ↑ampl[+ & -]'x = maxlog(x) - minlog(z)
+     ; ↓ampl[+ & -]'x = minlog(x) - maxlog(z)
+     ; --------------
+     ; Γ[+ & -]'y     = |y/(x+y)| & |-y/(x-y)|
+     ; ↑ampl[+ & -]'y = maxlog(y) - minlog(z)
+     ; ↓ampl[+ & -]'y = minlog(y) - maxlog(z)
      (define x (first srcs))
      (define y (second srcs))
 
-     (list (- (maxlog x) (minlog z)) ; exponent per x
-           (- (maxlog y) (minlog z)))] ; exponent per y
+     (if (*lower-bound-early-stopping*)
```

Comment:
Is this for performance? Don't compute the lower bound if not needed?

Reply:
Yes